app/src/main/java/org/mozilla/focus/fragment/HomeFragment.kt

```diff
@@ -172,19 +175,35 @@ private class HomeTileAdapter(
 private fun onBindBundledHomeTile(holder: TileViewHolder, tile: BundledHomeTile) = with (holder) {
     val bitmap = BundledTilesManager.getInstance(itemView.context).loadImageFromPath(itemView.context, tile.imagePath)
     iconView.setImageBitmap(bitmap)
+
+    titleView.text = tile.title
 }
 
 private fun onBindCustomHomeTile(uiLifecycleCancelJob: Job, holder: TileViewHolder, item: CustomHomeTile) = with (holder) {
-    launch(uiLifecycleCancelJob + CommonPool) {
-        val screenshot = HomeTileScreenshotStore.read(itemView.context, item.id) // TODO: if null, provide placeholder.
-        launch(UI) {
-            // Animate to avoid pop-in due to thread hand-offs. TODO: animation is janky.
-            ObjectAnimator.ofInt(iconView, "imageAlpha", 0, 255).apply {
-                interpolator = CUSTOM_TILE_ICON_INTERPOLATOR
-                duration = CUSTOM_TILE_ICON_TO_SHOW_MILLIS
-            }.start()
-            iconView.setImageBitmap(screenshot)
+    launch(uiLifecycleCancelJob + UI, CoroutineStart.UNDISPATCHED) {
```

Comment:
Is `launch` okay here, rather than `runBlocking`? I'm not too familiar with this but I guess the UI thread (where `onBindCustomHomeTile` lives isn't going to die) Just reading through the docs.

Reply:
Also why do you use a separate pool of threads?

Reply:
The two coroutines builders that are common are `launch` and `async`. At the top-level, they basically do the same thing, except for their return types (Job and Deferred respectively, the latter which can return a value).

`runBlocking` is idiomatically only used for tests – instead of handing the thread to someone else when the function is suspended, it blocks the thread, which ensures the outer function doesn't return (e.g. in the case of tests where the test completes when the function returns). 

The value in coroutines is of not blocking the thread you're running on but still being able to write intuitive concurrent code (i.e. it's written as if it's sequential) so it generally doesn't make sense to use `runBlocking` in production code (and the IDE will warn you too!).

```kotlin
launch(UI) {
    Log.d("lol", "a")
    delay(2000)
    Log.d("lol", "b")
}
```
Here the coroutine will suspend at the delay, allowing the UI thread to handle other tasks (like responding to touch events).

```kotlin
runBlocking(Ui) {
    Log.d("...
    delay(2000)
    Log.d("...
}
```

Here the coroutine will suspend at delay, *blocking* the UI thread from handling other tasks and probably locking up your UI thread. It's equivalent to not launching a coroutine and just calling `Thread.sleep(2000)`.

Reply:
wrt threading, when using coroutines, you can specify: 1) UI thread, 2) a specific executor, or 3) the common thread pool shared by the app.

If you're not running on the UI thread, it's generally best to put it on a common thread pool so that 1) we're not wasting resources by spawning and destroying new threads or 2) hoarding memory by keeping an executor just for this task.

In this case, we spawn two jobs without side effects (kinda – reading from disk might technically be a side effect but we lock and are not mutating state) so we can put them on two separate threads to take advantage of the concurrency, wait for them to complete, and perform the updated operations on the UI thread.

---

One problem comes from this though: putting tasks on a thread pool (as opposed to a single thread) creates concurrency – I'll spare you the details (and my wrists) on how to fix that until you need them. :)